webrender/src/frame_builder.rs

```diff
@@ -106,34 +106,123 @@ pub struct FrameBuilderConfig {
+    fn push_text(&mut self,
+                 run: &TextRun,
+                 local_rect: &LayerRect) {
+        // The combined local rect of the text-shadow primitive, is
+        // the union of all text runs, expanded by the blur radius
+        // for each run.
         let shadow_rect = local_rect.inflate(self.shadow.blur_radius,
                                              self.shadow.blur_radius);
         self.local_rect = self.local_rect.union(&shadow_rect);
+
+        // The blur shaders expect the source items to be placed in the
+        // center of the target (so offset by the blur radius). Since
+        // we are rendering glyphs into a target, we need to subtract the
+        // origin of their local bounding rect, to give a normalized
+        // glyph position.
+        let offset = LayerPoint::new(self.shadow.blur_radius - local_rect.origin.x,
+                                     self.shadow.blur_radius - local_rect.origin.y);
+
+        self.runs.push(ShadowTextRun {
+            run: run.clone(),
+            offset: offset,
+        });
+    }
+}
+
+// Maintains state when processing a text-shadow
+// stack, via push_text_shadow() and pop_text_shadow().
+struct TextShadowBuilder {
+    pending_shadows: Vec<PendingTextShadow>,
+    pending_texts: Vec<PendingTextRun>,
+}
+
+impl TextShadowBuilder {
+    fn new() -> TextShadowBuilder {
+        TextShadowBuilder {
+            pending_shadows: Vec::new(),
+            pending_texts: Vec::new(),
+        }
+    }
+
+    fn recycle(self) -> TextShadowBuilder {
+        TextShadowBuilder {
+            pending_shadows: recycle_vec(self.pending_shadows),
+            pending_texts: recycle_vec(self.pending_texts),
+        }
+    }
+
+    fn has_shadows(&self) -> bool {
+        !self.pending_shadows.is_empty()
+    }
+
+    // Add a new shadow to the list.
+    fn push_shadow(&mut self,
+                   shadow: TextShadow,
+                   clip_and_scroll: ClipAndScrollInfo,
+                   local_clip: LocalClip) {
+        self.pending_shadows.push(PendingTextShadow {
+            runs: Vec::new(),
+            shadow,
+            local_rect: LayerRect::zero(),
+            clip_and_scroll,
+            local_clip,
+        });
+    }
+
+    // Add a text run inside a push/pop text shadow stack.
+    fn push_text(&mut self,
+                 prim: TextRunPrimitiveCpu,
+                 local_rect: LayerRect,
+                 local_clip: LocalClip,
+                 clip_and_scroll: ClipAndScrollInfo) {
+        debug_assert!(self.has_shadows());
+
+        // Add the text run to each shadow in the list.
+        // TODO(gw): In the future we may be able to optimize
+        //           this to share glyph resources better.
+        for pending_shadow in &mut self.pending_shadows {
+            pending_shadow.push_text(&prim.run, &local_rect);
+        }
+
+        // If the color of the text run is not transparent,
```

Comment:
probably needs to be clarified as "not fully transparent", because alpha=0.5 is transparent and the condition below handles it.